rewrite-java/src/main/java/org/openrewrite/java/ReplaceAnnotation.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2024 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java;
+
+import lombok.EqualsAndHashCode;
+import lombok.Value;
+import org.openrewrite.*;
+import org.openrewrite.java.search.UsesType;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.JavaType;
+import org.openrewrite.java.tree.TypeUtils;
+
+@Value
+@EqualsAndHashCode(callSuper = false)
+public class ReplaceAnnotation extends Recipe {
+    @Option(displayName = "Annotation pattern to replace",
+            description = "An annotation pattern, expressed as a method pattern to replace.",
+            example = "@org.jetbrains.annotations.NotNull(\"Test\")")
+    String annotationPatternToReplace;
+    @Option(displayName = "Annotation template to insert",
+            description = "An annotation template to add instead of original one, will be parsed with `JavaTemplate`.",
+            example = "@NonNull")
+    String annotationTemplateToInsert;
+    @Option(displayName = "Type of inserted Annotation",
+            description = "The fully qualified class name of the annotation to insert.",
+            example = "lombok.NonNull")
+    String annotationFQN;
+    @Option(displayName = "Templates Artifact id",
+            description = "The Maven artifactId to load the inserted annotations type from, defaults to JDK internals.",
```

Comment:
We might have to add what classpath this gets loaded from, as that might not be clear to the user.

I am in fact wondering how we can get that to work unless the recipe is being used by another recipe library which provides the dependency.

Reply:
It would be interesting if the dependency could be demand-loaded using the Maven settings. Not sure if that is feasible or desirable. What is your take on this, @timtebeek ? We could also maintain (and document) that the recipe either has to be called from a recipe jar, which itself supplies the dependency or that the result will end up containing missing type attribution. But I am not sure if `ShortenFullyQualifiedTypeReferences` is going to like that.

Reply:
wouldn't it be sufficient if the project to migrate defines the needed dependency? As I understood, if defined in the dependency management the artifact would be on the classpath and all `JavaParser` are happy.

Reply:
Normally those classpath entries won't be available. I am assuming an exception is when the recipe is listed in the rewrite.yml file of the project where you run the migration, but I am a bit unsure about that.

Reply:
Tried it, it works in the tests without defining the artifact explicitly.